### PR TITLE
Reduce graph size through writing indexes directly into graph for ``map_blocks``

### LINKED ISSUE #8409: Task graphs on `.map_blocks` with many chunks can be huge
### What happened?

I'm getting task graphs > 1GB, I think possibly because the full indexes are being included in every task?

### What did you expect to happen?

Only the relevant sections of the index would be included

### Minimal Complete Verifiable Example

```Python
da = xr.tutorial.load_dataset('air_temperature')

# Dropping the index doesn't generally matter that much...

len(cloudpickle.dumps(da.chunk(lat=1, lon=1)))
# 15569320

len(cloudpickle.dumps(da.chunk().drop_vars(da.indexes)))
# 15477313

# But with `.map_blocks`, it really matters — it's really big with the indexes, and the same size without:


len(cloudpickle.dumps(da.chunk(lat=1, lon=1).map_blocks(lambda x: x)))
# 79307120

len(cloudpickle.dumps(da.chunk(lat=1, lon=1).drop_vars(da.indexes).map_blocks(lambda x: x)))
# 16016173
```

### MVCE confirmation

- [X] Minimal example — the example is as focused as reasonably possible to demonstrate the underlying issue in xarray.
- [X] Complete example — the example is self-contained, including all data and the text of any traceback.
- [X] Verifiable example — the example copy & pastes into an IPython prompt or [Binder notebook](https://mybinder.org/v2/gh/pydata/xarray/main?urlpath=lab/tree/doc/examples/blank_template.ipynb), returning the result.
- [X] New issue — a search of GitHub Issues suggests this is not a duplicate.
- [X] Recent environment — the issue occurs with the latest version of xarray and its dependencies.

### Relevant log output

_No response_

### Anything else we need to know?

_No response_

### Environment

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.9.18 (main, Aug 24 2023, 21:19:58)
[Clang 14.0.3 (clang-1403.0.22.14.1)]
python-bits: 64
OS: Darwin
OS-release: 22.6.0
machine: arm64
processor: arm
byteorder: little
LC_ALL: en_US.UTF-8
LANG: None
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.12.2
libnetcdf: None

xarray: 2023.10.1
pandas: 2.1.1
numpy: 1.26.1
scipy: 1.11.1
netCDF4: None
pydap: None
h5netcdf: 1.1.0
h5py: 3.8.0
Nio: None
zarr: 2.16.0
cftime: 1.6.2
nc_time_axis: None
PseudoNetCDF: None
iris: None
bottleneck: 1.3.7
dask: 2023.5.0
distributed: 2023.5.0
matplotlib: 3.6.0
cartopy: None
seaborn: 0.12.2
numbagg: 0.6.0
fsspec: 2022.8.2
cupy: None
pint: 0.22
sparse: 0.14.0
flox: 0.7.2
numpy_groupies: 0.9.22
setuptools: 68.1.2
pip: 23.2.1
conda: None
pytest: 7.4.0
mypy: 1.6.1
IPython: 8.14.0
sphinx: 5.2.1

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] Closes #8409
- [x] Tests added

When looking at then ``map_blocks`` variant of the benchmark [here](https://github.com/coiled/benchmarks/blob/main/tests/geospatial/test_climatology.py), I noticed that we had a 30 MiB graph for the medium variant. 10MiB of those were just the repeated adding of the PandasIndexes as an argument for map_blocks. Writing them directly to the graph will de-duplicate the value, and thus only have this object once instead of many many times. We can then reference the key for the function arguments.

The tokenise adds some overhead, so there is a drawback of this.

Happy to open an issue if required

Is this something that you all would consider merging?

cc @dcherian

### CODE REVIEW

**Comment by dcherian** [comment]
> Is this something that you all would consider merging?

Yes. This would extend what was done in https://github.com/pydata/xarray/pull/8412.

I *think* I initially chose this approach to make the graph absolutely embarrassingly parallel.

Can you add a test or confirm that dask has a test that makes sure the ordering scheduling is "nice" for this kind of case (widely duplicated dependency)?

**Comment by phofl** [comment]
> Can you add a test to dask or confirm that dask has a test that makes sure the ordering/scheduling is "nice" for this kind of case (widely duplicated dependency)?

This is the same pattern as we have with the open_dataset tasks and in the new dask array take implementation, we have existing tests covering these 2 patterns already

File: xarray/core/parallel.py [hunk 1]
```diff
@@ -551,6 +552,19 @@ def _wrapper(
             for isxr, arg in zip(is_xarray, npargs, strict=True)
         ]
 
+        indexes = {
```
Comment (dcherian) [inline xarray/core/parallel.py:555]:
```suggestion
        # only include new or modified indexes to minimize duplication of data
        indexes = {
```

**Comment by phofl** [comment]
> Ya this pattern used to suck, so I avoided it. But good to know it's fixed. Thanks for the PR!

TLDR is: If your task does not have a callable, it will be ignored during ordering
